### PR TITLE
config/main: Re-write config files - add to new config v3

### PR DESCRIPTION
**Pull request by harshavardhana**
New config format.

```
{
    "version": "3",
    "address": ":9000",
    "credential": {
        "accessKey": "WLGDGYAQYIGI833EV05A",
        "secretKey": "BYvgJM101sHngl2uzjXS/OBF/aMxAN06JrJ3qJlF"
    },
    "region": "us-east-1",
    "logger": {
        "file": {
            "enable": false,
            "fileName": "",
            "level": "error"
        },
        "syslog": {
            "enable": false,
            "address": "",
            "level": "debug"
        },
        "console": {
            "enable": true,
            "level": "fatal"
        }
    }
}
```

New command lines in lieu of supporting XL.

Minio initialize filesystem backend.

```
$ minio init fs <path>
```

Minio initialize XL backend.

```
$ minio init xl <url1>...<url16>
```

For 'fs' backend it starts the server.

```
$ minio server
```

For 'xl' backend it waits for servers to join.

```
$ minio server
... [PROGRESS BAR] of servers connecting
```

Now on other servers execute 'join' and they connect.

```
....
minio join <url1> -- from <url2> && minio server
minio join <url1> -- from <url3> && minio server
...
...
minio join <url1> -- from <url16> && minio server
```

### CODE REVIEW

**Comment by harshavardhana** [comment]
Build still fails, fixing it .. - keep it here for reviews

**Comment by abperiasamy** [comment]
* address cannot be empty. let the default be 9000. Default config file should be usable. 
* Console section is missing in the log file. Console should be enabled by default. 
* We cannot have default values like nul. Log level should have default value filled in.

File: config-logger-main.go [hunk 1]
```diff
@@ -34,130 +38,223 @@ var configLoggerCmd = cli.Command{
 USAGE:
    minio config {{.Name}} OPERATION [ARGS...]
 
-   OPERATION = add | list | remove
+OPERATION:
+   enable  LOGGER mongo [ADDRESS] [LOG-LEVEL,..]
+   enable  LOGGER file [FILENAME] [LOG-LEVEL,..]
+   disable LOGGER
+   list
```
Comment (abperiasamy) [inline config-logger-main.go]: Add a section for LOG-LEVEL: and describe its possible values.

Comment (abperiasamy) [inline config-logger-main.go]: Also rename LOG-LEVEL to LEVEL. It takes only one value.

File: config-logger-main.go [hunk 2]
```diff
@@ -34,130 +38,223 @@ var configLoggerCmd = cli.Command{
...
+      $ minio config {{.Name}} enable syslog localhost:5514 DEBUG,CRITICAL
 
-   4. List currently configured logger.
+   3. List of all currently configured loggers.
       $ minio config {{.Name}} list
 
-   5. Remove/Reset a configured logger.
-      $ minio config {{.Name}} remove mongo
+   4. Remove/Reset a configured logger.
+      $ minio config {{.Name}} disable mongo
```
Comment (abperiasamy) [inline config-logger-main.go]: Examples also needs to be updated for LEVEL value.

File: config-logger-main.go [hunk 3]
```diff
@@ -34,130 +38,223 @@ var configLoggerCmd = cli.Command{
...
+	address := args.Get(0)
+	if address == "" {
+		fatalIf(probe.NewError(errDummy), "MongoDB logger cannot be enabled for empty address.", nil)
+	}
+	_, _, e := net.SplitHostPort(address)
+	if e != nil {
+		fatalIf(probe.NewError(e), "MongoDB endpoint address is of invalid format "+address, nil)
+	}
+
+	logLevels := strings.Split(args.Get(1), ",")
```
Comment (abperiasamy) [inline config-logger-main.go]: level is  now a single value string.

File: config-logger-main.go [hunk 4]
```diff
@@ -34,130 +38,223 @@ var configLoggerCmd = cli.Command{
...
+	}
+
+	// Load current config.
+	conf, err := loadConfig()
+	fatalIf(err.Trace(), "Unable to load config.", nil)
+
+	conf.Logger.Mongo = mongoLogger{
+		Enable:    true,
+		Addr:      address,
+		LogLevels: logLevels,
```
Comment (abperiasamy) [inline config-logger-main.go]: LogLevels -> Level

File: config-logger-main.go [hunk 5]
```diff
@@ -34,130 +38,223 @@ var configLoggerCmd = cli.Command{
...
+   4. Remove/Reset a configured logger.
+      $ minio config {{.Name}} disable mongo
 `,
 }
 
-// Inherit at one place
-type config struct {
-	*configV2
+// enable MongoDB logging.
+func enableMongoLogger(args cli.Args) {
```
Comment (abperiasamy) [inline config-logger-main.go]: Rewrite this code to be modular and cleaner. Separate the parsing logic out.
-> func logParseMongo(options []string) logMongo
-> func logParseConsoleoptions []string) logConsole
-> func logParseSyslog(options []string) logSyslog
...
Parse function converts CLI args to typed structure.

Then rest of the code will fit in neatly and become generic.

**Comment by harshavardhana** [comment]
> -> func enableMongoLogger(options []string)

The reason to avoid this is cli.Args implements better helpers, i don't see why 'options []string' will provide   any benefit. 'cli.Args' is also more readily readable.

File: server.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		apiServer.TLSConfig.Certificates = make([]tls.Certificate, 1)
+		apiServer.TLSConfig.Certificates[0], err = tls.LoadX509KeyPair(srvConfig.TLSConfig.CertFile, srvConfig.TLSConfig.KeyFile)
+		if err != nil {
+			return nil, probe.NewError(err)
+		}
+	}
+	// Minio server config
+	apiServer.Addr = srvConfig.Addr
+	apiServer.Handler = apiHandler
+	apiServer.MaxHeaderBytes = 1 << 20
```
Comment (abperiasamy) [inline server.go]: 1MB is the default. You dont need to set it explicitly. Even otherwise, it should always require a comment like // 1 MB

**Comment by harshavardhana** [comment]
> logLevels -> level - It only takes one value.

why only one value, at a time we can only configure 4 different log levels . What if one wishes to log all levels to a file or syslog? - Single level might not be correct from a logging point of view.

File: logger-mongo-hook.go [hunk 7]
```diff
@@ -25,31 +25,29 @@ import (
...
 )
 
+type mongoLogger struct {
+	Enable    bool     `json:"enable"`
+	Addr      string   `json:"address"`
+	LogLevels []string `json:"logLevels"`
+}
+
 // mongoDB collection
 type mongoDB struct {
```
Comment (abperiasamy) [inline logger-mongo-hook.go]: All these code has to be rewritten completely. We should never have structs like mongoDB and member named c

File: logger-syslog-hook_nix.go [hunk 8]
```diff
@@ -33,8 +39,8 @@ type syslogHook struct {
 	syslogRaddr   string
 }
 
-func log2Syslog(network, raddr string) *probe.Error {
-	syslogHook, e := newSyslog(network, raddr, syslog.LOG_ERR, "MINIO")
+func log2Syslog(raddr string) *probe.Error {
```
Comment (abperiasamy) [inline logger-syslog-hook_nix.go]: Code is incomplete and also no comments.

File: main.go [hunk 9]
```diff
@@ -163,11 +169,33 @@ func checkMainSyntax(c *cli.Context) {
...
+	if conf.Logger.Syslog.Enable {
+		err := log2Syslog(conf.Logger.Syslog.Addr)
+		fatalIf(err.Trace(conf.Logger.Syslog.Addr), "Unable to initialize syslog logger hook.", nil)
+	}
+	if conf.Logger.File.Enable {
+		err := log2File(conf.Logger.File.Filename)
+		fatalIf(err.Trace(conf.Logger.File.Filename), "Unable to initialize file logger hook.", nil)
+	}
+}
+
```
Comment (abperiasamy) [inline main.go:174]: This code should be rewritten.

**Comment by abperiasamy** [comment]
Resubmit the patch with just config changes. Logging can be brought in separately. Config entries will be NOOP. Use standard golang logger to console for now.

**Comment by harshavardhana** [comment]
> Resubmit the patch with just config changes. Logging can be brought in separately. Config entries will be NOOP. Use standard golang logger to console for now.

It can be fixed in the same patch, since it is part of config changes. Let me do that.

**Comment by abperiasamy** [comment]
level value should be lower case.

File: config-logger-main.go [hunk 10]
```diff
@@ -34,130 +36,215 @@ var configLoggerCmd = cli.Command{
...
+// parse Console logger type.
+func parseConsole(logLevel string) (consoleLogger, *probe.Error) {
+	// Validate input log level.
+	if !isValidLogLevel(logLevel) {
+		return consoleLogger{}, probe.NewError(errors.New("Unrecognized log level " + logLevel))
 	}
-	conf, err := loadConfigV2()
-	fatalIf(err.Trace(), "Unable to load config", nil)
+	// Initialize console logger.
+	clogger := consoleLogger{
```
Comment (abperiasamy) [inline config-logger-main.go]: conLogger

File: access-keys.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Minio Cloud Storage, (C) 2015 Minio, Inc.
```
Comment (abperiasamy) [inline access-keys.go]: 2016

Comment (abperiasamy) [inline access-keys.go]: also rename file to access-key.go

File: config-logger-main.go [hunk 12]
```diff
@@ -34,130 +36,215 @@ var configLoggerCmd = cli.Command{
 USAGE:
    minio config {{.Name}} OPERATION [ARGS...]
 
-   OPERATION = add | list | remove
+OPERATION:
+   enable  LOGGER mongo   [ADDRESS]  [LEVEL]
+   enable  LOGGER file    [FILENAME] [LEVEL]
```
Comment (abperiasamy) [inline config-logger-main.go]: LOGGER -> log

**Pull request merged**